Dome training material for AI school. (#5907)
galaxyproject/training-material: topics/statistics/tutorials/dome/tutorial.md

contributions:
  authorship:
  - fpsom
  - sfragkoul
tags:
- elixir
- ai-ml
priority: 6
notebook:
  language: python
  pyolite: true
---
[//]: # (From Docs: Introduction)
## The need for standardization

With the significant drop in the cost of many high-throughput technologies, vast amounts of biological data are being generated and made available to researchers. 
Machine learning (ML) has emerged as a powerful tool for analyzing data related to cellular processes, genomics, proteomics, post-translational modifications, metabolism, and drug discovery, offering the potential for transformative medical advancements. 
This trend is evident in the growing number of ML publications, showcasing a wide array of modeling techniques in biology. 
However, although ML methods should ideally be experimentally validated, this occurs in only a small portion of the studies. 
We believe the time is right for the ML community to establish standards for reporting ML-based analyses to facilitate critical evaluation and enhance reproducibility.[@DOME]

> ![Bar plots showing the trend of usage of ML in Biology.](https://github.com/BiodataAnalysisGroup/training-material/blob/dome-training/topics/statistics/images/dome/ML_Biology.png)

Reviewer: could you add the image to the GTN directly?

> ![Bar plots showing the trend of usage of ML in Biology.](https://github.com/BiodataAnalysisGroup/training-material/blob/dome-training/topics/statistics/images/dome/ML_Biology.png)

Reviewer: To point to internal images, you can use this method:

```suggestion
![Bar plots showing the trend of usage of ML in Biology.]({% link topics/statistics/images/dome/ML_Biology.png %})
```

The ability of a model to maintain its performance on new data is referred to as its generalization power. 
Achieving strong generalization is a key challenge in developing ML models; without it, trained models cannot be effectively reused. 
Properly preprocessing data and using it in an informed way are essential steps to ensure good generalization.

> **Note: Further Reading**  
> State-of-the-art ML models are often capable of memorizing all variations within the training data. As a result, when evaluated on data from the training set, they may give the false impression of excelling at the given task. However, their performance tends to diminish when tested on independent data (called the test or validation set), revealing a lower generalization power. To address this issue, the original dataset should be randomly split into non-overlapping parts. The simplest method involves creating separate training and test sets (with a possible third validation set). Alternatively, more robust techniques like cross-validation or bootstrapping, which repeatedly create different training/testing splits from the available data, are often preferred.
	

> 	Handling overlap between training and test data can be especially challenging in biology. For instance, in predicting entire gene or protein sequences, ensuring data independence might require reducing homologs in the dataset. In modeling enhancer–promoter contacts, a different criterion may be needed, such as ensuring that no endpoint is shared between training and test sets. Similarly, modeling protein domains may require splitting multidomain sequences into their individual domains before applying homology reduction. Each biological field has its own methods for managing overlapping data, making it crucial to consult prior literature when developing an approach.

Reviewer: Having these indents at the start of a sentence turns it into a code block, which I don't think you wanted.

**Key Questions**

- Does the model use data from __other__ ML algorithms as input? 
- If yes, which ones? 
- Is it clear that training data of initial predictors and meta-predictor are __independent__ of test data for the meta-predictor?

> **From Example Publication**  
> Yes, predictor output is a binary prediction computed from the consensus of other methods; Independence of training sets of other methods with test set of meta-predictor was not tested since datasets from other methods were not available.

## 2.3 Data encoding

Data encoding is the process of transforming data from one format or structure into another, often to make it easier for ML models or computational systems to process. 
In ML, data often needs to be encoded to ensure that it can be effectively interpreted by algorithms, especially for algorithms that require numerical input (e.g., neural networks, SVMs).

> **Key Questions**
> 
> - How were the data __encoded__ and __preprocessed__ for the ML algorithm?
> 
> > **From Example Publication**  
> > Label-wise average of 8 binary predictions.

Reviewer: It would be great if you could use GTN-style question boxes for these Key Question sections, so something like:

```suggestion
> <question-title> Key Questions </question-title>
>
> 1. How were the data __encoded__ and __preprocessed__ for the ML algorithm?
>
> > <solution-title></solution-title>
> > 1. **From Example Publication**  
> >    Label-wise average of 8 binary predictions.
> >
> {: .solution}
{: .question}
```

- If yes, __which__ ones?

> **From Example Publication**  
> No. 

## 2.8 Availability of configuration

Availability of configuration refers to the accessibility and transparency of the settings, parameters, and options that can be adjusted or customized in a ML model or system. 
These configurations control how the model is trained, how it makes predictions, and how it operates in different environments. 
Ensuring that the configuration is available, flexible, and easy to modify is important for reproducibility, fine-tuning, and deployment of models.

**Key Questions**

- Are the hyperparameter configurations, optimization schedule, model files and optimization parameters __reported__? 
- If yes, __where__ (for example, URL) and __how__ (license)?

> **From Example Publication**  
> Not applicable.

   

[//]: # (From Docs: Chapter 03)

Good overall performance and the model's ability to generalize well to unseen data are crucial factors that significantly impact the applicability of any proposed ML research. However, several other important aspects related to ML models must also be considered.

> > **Note: Further Reading**  

Reviewer: further reading could be in a comment box, or a details box (collapsed by default)

```
> <comment-title>Further Reading</comment-title>
> your text here..
{: .comment}
```

(replace comment with details if you want the box collapsed by default)

{: .comment}
	

## 1.1 Provenance 

Provenance of data refers to the origin, history, and lineage of data—essentially, tracking where the data came from, how it has been processed, and how it has moved through various systems. It’s like a detailed record that traces the data's life cycle from creation to its current state. Understanding data provenance helps ensure transparency, trustworthiness, and reliability in data usage.

> <question-title> Key Questions </question-title>
>
>  What is the __source__ of the data (database, publication, direct experiment)? 
> If data are in __classes__, how many data points are available in each class—for example, total for the positive (*Npos*) and negative (*Nneg*) cases?
> If __regression__, how many real value points are there? 
> Has the dataset been __previously used__ by other papers and/or is it recognized by the community?
>
> > <solution-title></solution-title>

> > >**From Example Publication**  

Reviewer 2: Where is this example publication? Please ignore if there is already a link provided. Additionally, there should be short summary of the publication in the tutorial.

Author: This is already cited in the text. A summary of the publication was added.

> <question-title> Key Questions </question-title>
>
> How __many data points__ are in the training and test sets?  
> Was a __separate validation__ set used, and if yes, how large was it?  
> Are the __distributions__ of data types (*Npos* and *Nneg*) in the training and test sets different? Are the distributions of data types in both training and test sets plotted?
>
> > <solution-title></solution-title>
> >**From Example Publication**  
> > training set: N/A. *Npos,test* = 339,603 residues. *Nneg,test* = 6,168,717 residues. No validation set. 5.22% positives on the test set.
> {: .solution}
{: .question}
      

## 1.3 Redundancy between data splits

 Redundancy between data splits occurs when the same data points are present in more than one of the training, validation, or test sets. This is undesirable because it can distort model evaluation and lead to overoptimistic performance metrics (e.g. eliminating data points more similar than X%). 

> This may effect the mfodel by introcuding an overfitting risk, unreliable performance metrics and/or lack of generalization.

Reviewer 2: ```suggestion
This may impact the model by introducing the risk of overfitting and poor generalization rendering the performance metrics as unreliable.
```

For example neural networks are often criticized for being "black boxes," as their internal workings (like hidden layers) are less transparent, making them more difficult to trust.
There are generally two types of interpretability:

- Global Interpretability refers to the ability to understand the overall behavior of the model across all predictions. It involves explaining how the model works as a whole and what features are important in determining predictions. 

- Local interpretability focuses on understanding individual predictions made by the model. It aims to explain why a specific input led to a particular output.

> <question-title> Key Questions </question-title>
>
>  Is the model __black box__ or __interpretable__? 
> If the model is interpretable, can you give clear __examples__ of this?
>
> > <solution-title></solution-title>
> >**From Example Publication**  
> > Transparent, in so far as meta-prediction is concerned. Consensus and post processing over other methods predictions (which are mostly black boxes). No attempt was made to make the meta-prediction a black box.
> {: .solution}
{: .question}

## 3.2 Output

> The output of a machine learning model refers to the predictions or classifications made by the model after processing input data. 

Reviewer 2: ```suggestion
The output of a machine learning model refers to the predictions made by the model on unseen data after learning on input data. 
```

Author: I believe this adjustment may subtly alter the intended meaning of the text.

-   __Multi-label classification__ includes the assignment of multiple classes to a single input.  

> <question-title> Key Questions </question-title>
>
>  Is the model __classification__ or __regression__?
>
> > <solution-title></solution-title>
> >**From Example Publication**  
> > Classification, i.e. residues thought to be disordered.
> {: .solution}
{: .question}

## 3.3 Execution time

 
Execution time in the context of ML refers to the duration it takes for a model to perform a specific task, such as training, predicting, or evaluating performance. 
Understanding and measuring execution time is crucial for various reasons, including optimizing model performance, resource management, and user experience. 
  

CPU time of single representative execution on standard hardware (e.g. seconds on desktop PC).

> <question-title> Key Questions </question-title>
>
> How much __time__ does a single representative prediction require on a standard machine (for example, seconds on a desktop PC or high-performance computing cluster)?
>
> > <solution-title></solution-title>
> >**From Example Publication**  

> > > ca. 1 second per representative on a desktop PC.

Reviewer 2: ```suggestion
> > Approximately 1 second per representative on a desktop PC.
```

{: .question}

## 4.3 Comparison

 
Comparison typically refers to the evaluation of different models, algorithms, or configurations to identify which one performs best for a specific task. 
This process is essential for selecting the most suitable approach for a given problem, optimizing performance, and understanding the strengths and weaknesses of various methods. 
  
> <question-title> Key Questions </question-title>
>
>  Was a comparison to __publicly available__ methods performed on benchmark datasets? 
> Was a comparison to __simpler baselines__ performed?
> 
> > <solution-title></solution-title>
> >**From Example Publication**  
> > DisEmbl-465, DisEmbl-HL, ESpritz Disprot, ESpritz NMR, ESpritz Xray, Globplot, IUPred long, IUPred short, VSL2b. Chosen methods are the methods from which the meta prediction is obtained.
> {: .solution}
{: .question}

## 4.4 Confidence

 
Confidence in the context of ML refers to the measure of certainty or belief that a model's prediction is accurate. 
It quantifies the model's certainty regarding its output, which is particularly important in classification tasks, where decisions need to be made based on predicted class probabilities.

> This can be supported with medthods such as confidence intervals and statistical significance.

Reviewer 2: ```suggestion
This can be supported by methods such as confidence intervals and statistical significance.
```
